Add configmap and secret count metrics

wenzstev opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

This pull request adds two additional metrics to the Kubernetes State check:

- `kubernetes_state.secret.count`
- `kubernetes_state.configmap.count`

These metrics count the number of secrets and configmaps respectively, and can be aggregated by the `namespace` tag. These metrics can be toggled with two new settings for the check:

- `enable_configmap_count`
- `enable_secret_count`

These default to false; the metrics won't be gathered unless they're set to true. The metrics are gathered in the same way as other aggregation metrics.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

Jira card [here](https://datadoghq.atlassian.net/browse/CONT-3648?atlOrigin=eyJpIjoiN2JjMjBiYTMxMTgwNGJlMzlmYTUwYmRiYzUyNjc0ZjUiLCJwIjoiaiJ9). We have counts for other Kubernetes resources, but not these. 

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

The code currently checks for the presence of the `enable` values after potential aggregate metrics are found. There may be a better way to do this. 

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

1. Spin up a version of the agent with this check using Helm.
2. Manually edit the Cluster Agent's `ClusterRole` that is generated in order to ensure that the agent has permission to list `ConfigMap` and `Secret` resources. 
3. Confirm that the metric count matches the actual count of the given resource. 

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

hush-hush committed:
Starts splitting config helpers from the config package

As we continue to migrate the Agent code to components the many helpers
in pkg/config need to be moved somewhere else. The goal is to have a
simple API for the Config component while offering helpers on top of it.

The goal is to have helpers that can take either a "pkg/config:Config" or a
"comp/core/config:Component". This will facilitate the migration to
components.

For now the helpers are moved to pkg/config/utils. This is temporary as
we brainstorm the best place for them. Starting to split them from the
main config package is still useful.

hush-hush committed:
Fix dynamic secrets scrubbing in flare

The flare now request a new scubber each time. We need to propagate the
dynamic replacers from secrets and config.

hush-hush committed:
Update documentation wording for 'exclude_ec2_tags'

nenadnoveljic committed:
default config fix (#16664)

* default config fix

* renamed file from default to example

* codeowners update

* codeowners update

* remove query metrics

guyarb committed:
usm: sowatcher: Fixed leak (#16650)

* usm: sowatcher: Fixed leak

We registered paths by the number of mappings, but cleanning them by the number of processes.
A single process might have multiple mappings of the same so, which makes those numbers different

* Fixed typo

* Fixed license

* Addresses CR comment

* Addresses CR comment

* Added task change

* Fixed script

* Fixed CI

* Fixed lint

* Addresses CR comment

* Fixed path

lebauce committed:
epforwarder: use stream strategy for protobuf payloads (#16635)

Commit:
Move the trigger_release stage (#16671)

We want the trigger_release stage to run when:
* The basic kitchen tests pass
* The packages are present in the dd-release-artifacts bucket

We want to prevent situations when e.g. one of the chocolatey jobs
fails and makes the trigger_release jobs to be skipped.

lebauce committed:
[CWS] Fix SBOM generation for container root filesystems (#16657)

* Put error in result channel

* Use different SBOM caches for core agent and system-probe

* Fix slow walk when using symlinked paths

* Disable host SBOM generation by default

safchain committed:
[CWS] fix interpreter inode (#16644)

L3n41c committed:
[ksm-core] Remove the `kube_namespace` tag from the CRD metrics (#16656)

TCheruy committed:
[SNMP] Set default snmp version empty string for snmp walk command (#16681)

* Set default snmp version to 2c for snmp walk command to avoid warning

* Default to empty version to better mimic SNMP integration

hush-hush committed:
Send 'agent_checks' payload 1 min after startup

The 'agent_checks' payload contains the status of every checks run by
the agent. It's sent every 10 min. We no longer wait 10 min for the first
payload to be sent. Instead we send the first payload after 1 min and
every 10 min after that.

Commit:
[rpm][deb] Fix setting owner on site-packages directories (#16684)

AliDatadog committed:
[SBOM] fix:remove cache cleaner in collector and duplicated telemetry metrics. Change default sbom options. (#16629)

just-chillin committed:
[process-agent] Log errors on failure to initialize (#16611)

* [process-agent] Log errors on failure to initialize

* Remove logger component

* fix misspelling

* Use logger by default and fall back to stdout if nil

hush-hush committed:
Adding timeout to the health command

hush-hush committed:
Remove duplicated Logger and Config name in Agent cmd

Similar to what the cluster agent does, we don't want to duplicate the
logger and config name in every command.

wenzstev committed:
merging changes onto branch

L3n41c committed:
Do not run SBOM collection while running `agent check` (#16688)

KSerrania committed:
[rpm] Fix wrongly set package permissions in posttrans (#16686)

ojung committed:
pkg/trace/api: Fix tag accumulation bug (#16690)

* Fix tags being accumulated across requests

* Truncate ddtags at 4000 characters

* Add release note

* Close servers in tests

* Remove release note

kangyili committed:
[Orchestrator] Check node collector if cluster collector is activated (#16692)

check node collector if cluster collector is activated

paulcacheux committed:
[CWS] fix handling of empty argv in module init (#16706)

* [CWS] fix handling of empty argv in module init

* add test

* fix go lint

truthbk committed:
[golang] bumping to 1.19.8 (#16633)

* [golang] bumping to 1.19.8

* [reno] adding release note

* [circleci] updating readme

* [gitlab] lint jobs need missing image suffix option added

* [circleci] bump image

* [gitlab] accidentally introduced typo

* [gitlab] bump windows build images too

* [gitlab] fix windows lint job with no suffix

* [gitlab] set final builder image tags

guyarb committed:
usm: Fixed kafka docker setup, after external change (#16714)

clarkb7 committed:
AllowClosedSource NG installer fixes (#16648)

The failing kitchen tests are unrelated to this PR.

Commit:
[kitchen] Temporarily remove Debian 9 from supported platforms (#16728)

kacper-murzyn committed:
Release date updated + cleanup (#16679) (#16722)

kacper-murzyn committed:
Update release.json and Go modules for 6/7.45.0-rc.3 (#16720)

paulcacheux committed:
flare: fix `security-agent.yaml` config file inclusion (#16718)

* flare: fix `security-agent.yaml` config file inclusion

* add release note

Kaderinho committed:
remove fips.https and fips.port_range_start from config_template.yaml (#16732)

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@epita.fr>

lebauce committed:
[CWS] Fix releasing of scoped variables (#16727)

* Fix releasing of scoped variables

* Add unit test on scoped variable releasing

tbavelier committed:
Fix labels/annotations as tags feature on APIService KSM metrics (#16737)

sgnn7 committed:
[config] Update available sites in the config template

The current site list is out of date. This change ensures that we
have all the available ones identified.

ahmed-mez committed:
pkg/clusteragent/admission: support inject all libs via RC (#16653)

guyarb committed:
Update system_probe.py (#16713)

* Update system_probe.py

* Fixed lint

guyarb committed:
usm: remove gotos (#16625)

* usm: remove gotos

* Fixed cr comment

* Fixed cr comment

guyarb committed:
usm: protocol classification: Classify in the kernel only if the feature is supported (#16663)

* usm: protocol classification: Classify in the kernel only if the feature is supported

* Fixed CI

guyarb committed:
usm: performance: Massive improvement in encoder (#16497)

* usm: performance: Massive improvement in encoder

Improving 70% in allocations, 60% in total bytes, and 50% in runtime"

* Fixed cr comments

* Using sync.pool

safchain committed:
[CWS] fix profiles path (#16725)

paulcacheux committed:
[CWS] fix build tags generation in accessors/field handlers generation code (#16654)

* fix generation of build tags after copyright and code generated line

* cleanup copyrights ignores

guyarb committed:
usm: Improve connection creation to reduce allocations (#16621)

just-chillin committed:
[process-agent] Fix bug where disabled message is not shown (#16703)

* [process-agent] Fix bug where disabled message is not shown

mftoure committed:
[CWS][SEC-4301] Add security context to pods in workload metadata store (#16669)

* [WIP] Retrieve security context on pod's containers

* [WIP] Retrieve security context on pod's containers

* Retrieve security context on pod's containers

just-chillin committed:
[process-agent] Change sysprobe config missing error to info on Darwin (#16745)

Do not log error message due to missing system-probe.yaml file on dawin

safchain committed:
[CWS] add option to disable path resolution (#16564)

ganeshkumarsv committed:
Update CodeQL Bundle version (#16640)

* Update codeql-analysis.yml

* remove codeql config

* Delete codeql_config.yml

AliDatadog committed:
[Container Filter] Fix metrics exclusion for corechecks (#16697)

julien-lebot committed:
Fix APM telemetry casing & order (#16765)

Kitchen failures strictly unrelated to this PR

paulcacheux committed:
trivy: revert to require `v0.0.0-00010101000000-000000000000` for trivy (#16766)

paulcacheux committed:
bump `github.com/avast/retry-go/v4` and `github.com/cenkalti/backoff/v4` (#16768)

* bump `github.com/avast/retry-go/v4` to v4.3.4

* bump `github.com/cenkalti/backoff/v4` to v4.2.1

gbbr committed:
pkg/trace/api: fix flaky TestFlatten test (#16515)

hmahmood committed:
[system-probe] Reorg ebpf code for reuse with CO-RE network tracer (#16157)

hmahmood committed:
Add config for disabling/enabling netlink conntracker fallback (#16647)

safchain committed:
[CWS] remove runtime-security folder (#16763)

ahmed-mez committed:
pkg/trace/api: fix test lint (#16775)

lebauce committed:
Update fakeintake README.md (#16658)

safchain committed:
[CWS] fix macos build (#16777)

clamoriniere committed:
fix: support scheme in EVP forwarder (#16750)

* fix: support scheme in EVP forwarder

* add release note

* Update releasenotes/notes/logs_config-support-url-in-dd_url-4e28e17d0b3ad6d0.yaml

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

---------

Co-authored-by: DeForest Richards <56796055+drichards-87@users.noreply.github.com>

wenzstev committed:
added release note

clamoriniere commented on pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:
```diff
@@ -138,6 +138,12 @@ type KSMConfig struct {
 	LeaderSkip bool `yaml:"skip_leader_election"`
 
+	//EnableConfigMapCount allows for enabling or disabling the kubernetes_state.configmap.count metric
+	EnableConfigMapCount bool `yaml:"enable_configmap_count"`
```
In addition I don't think we need these 2 variables. we should be able to trigger the creation of the metrics only if the `configmap` or `secret` is added of the list of monitored resources.

clamoriniere commented on pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:
```diff
@@ -503,6 +509,14 @@ func (k *KSMCheck) processMetrics(sender aggregator.Sender, metrics map[string][
 		for _, metricFamily := range metricsList {
 			// First check for aggregator, because the check use _labels metrics to aggregate values.
 			if aggregator, found := k.metricAggregators[metricFamily.Name]; found {
+				if metricFamily.Name == "kube_configmap_info" && !k.instance.EnableConfigMapCount {
+					log.Tracef("ConfigMap metric found but counting ConfigMaps is disabled.")
+					continue
+				}
+				if metricFamily.Name == "kube_secret_info" && !k.instance.EnableSecretCount {
+					log.Tracef("Secret metric found but counting Secrets is disabled.")
+					continue
+				}
```
imo, this is not needed if we remove the 2 options as I suggested in a previous comment.

clamoriniere reviewed (commented):
Add few comments.
the changes made in `pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go` should be enough.

Please also plan a pull request to add the new metrics in https://github.com/DataDog/integrations-core/blob/master/kubernetes_state_core/metadata.csv 🙇

wenzstev committed:
removed extraneous logic for the metrics

wenzstev commented:
@clamoriniere Done. I removed all of the additional code from the main check class. One just needs to add the given collectors to the Kubernetes State Core configuration in order to get the metrics. Also updated the Helm chart PR to reflect this: [https://github.com/DataDog/helm-charts/pull/1034](https://github.com/DataDog/helm-charts/pull/1034/files)

wenzstev committed:
renamed problematic release notes file

estherk15 reviewed (approved):
Looks good for docs

wenzstev committed:
shorten releasenote name

wenzstev committed:
Delete previous

wenzstev committed:
Update releasenotes/notes/added-two-new-metrics-to-KSM-core-check-df655446dc43cded.yaml

Co-authored-by: Cedric Lamoriniere <cedric.lamoriniere@datadoghq.com>
